### DISCUSSION TITLE
Apache server hang when calling similarity method

### DISCUSSION THREAD

**Discussion by leinad9905** [explosion/spaCy#11693] [Help: Coding & Implementations] [answered]
New to spaCy and wanting to match the user's utterance to a list of examples using similarity method, this runs the first time, but the server would hang the second time with another message, this is on an Apache server on a cloud platform with shared, non-dedicated resources if this info is useful. Sorry if this might not be the right place to ask.

```
intent = {'intent': '', 'similarity': 0.0, 'response': ''}
message_nlp = nlp(message)
for x in intent_basic_info:
    for y in x['message']:
        message_nlp_example = nlp(y.lower())
        similarity = message_nlp.similarity(message_nlp_example)
        if similarity > intent['similarity']:
            intent['intent'] = x['intent']
            intent['similarity'] = similarity
            intent['response'] = random.choice(x['response'])
```

spaCy version    3.4.2                         
Platform         Linux-5.8.0-50-generic-x86_64-with-glibc2.32
Python version   3.8.10                        
Pipelines        en_core_web_md (3.4.1)

**Comment by polm** [comment, marked as answer]
Sorry you're having trouble with this, but with the information you've given us this is pretty hard to debug. 

To be clear, is "another message" input, or do you get an error message? If you get no error message, can you kill the server to get a backtrace to show where it was stuck? Or can you add logging?

How many items are you iterating over here? How many items are in `intent_basic_info` and then in each of the `"message"` members? Calculating similarity in a nested for loop is generally very slow, and you should use something like [annoy](https://github.com/spotify/annoy) instead.

**Reply by leinad9905** [reply]
Hey polm, thanks for the response. Probably could have written this question more clearly, pretty cool tool you've got here. 

And yes, it take an input from the user and process it,  then loop through processed example messages to see if they are similar enough, using similarity method. The server hangs and there's no error message. Still new to everything will look more into server backtrace and logging.

There's a few dozen of items total it's iterating over (in intent_basic_info). Will check out annoy.

Thanks again, I guess just needed some pointers, this definitely helped

**Reply by polm** [reply]
OK, thanks for the details. If you do understand more from logging let us know if we can help.

If you have only a few dozen items, it does seem unlikely that would cause the server to hang. Using annoy or similar would be faster, and helpful if you have a lot of requests, but the cause of the hanging might be something else.
